vpr/src/route/rr_graph_fpga_interchange.cpp

```diff
@@ -0,0 +1,1706 @@
+        }
+
+        for (size_t i = 0; i < segment_inf.size(); ++i) {
+            wire_name_to_seg_idx_[segment_inf_[RRSegmentId(i)].name] = i;
+        }
+
+        auto primLib = ar_.getPrimLibs();
+        auto portList = primLib.getPortList();
+
+        for (auto cell_mapping : ar_.getCellBelMap()) {
+            size_t cell_name = cell_mapping.getCell();
+
+            int found_valid_prim = false;
+            for (auto primitive : primLib.getCellDecls()) {
+                bool is_prim = str(primitive.getLib()) == std::string("primitives");
+                bool is_cell = cell_name == primitive.getName();
+
+                bool has_inout = false;
+                for (auto port_idx : primitive.getPorts()) {
+                    auto port = portList[port_idx];
+
+                    if (port.getDir() == INOUT) {
+                        has_inout = true;
+                        break;
+                    }
+                }
+
+                if (is_prim && is_cell && !has_inout) {
+                    found_valid_prim = true;
+                    break;
+                }
+            }
+
+            if (!found_valid_prim)
+                continue;
+
+            for (auto common_pins : cell_mapping.getCommonPins()) {
+                for (auto site_type_entry : common_pins.getSiteTypes()) {
+                    for (auto bel : site_type_entry.getBels()) {
+                        bel_cell_mappings_.emplace(bel);
+                    }
+                }
+            }
+        }
+    }
+
+    void build_rr_graph() {
+        create_switch_inf();
+        create_tile_id_to_loc();
+        create_uniq_node_ids();
+        create_sink_source_nodes();
+        create_pips_();
+        process_nodes();
+#ifdef DEBUG
+        print_virtual();
+#endif
+        virtual_to_real_();
+        pack_to_rr_graph();
+#ifdef DEBUG
+        print();
+#endif
+        finish_rr_generation();
+    }
+
+  private:
+    Device::Reader& ar_;
+    DeviceContext& device_ctx_;
+    const vtr::vector<RRSegmentId, t_segment_inf>& segment_inf_;
+    const DeviceGrid& grid_;
+    const enum e_base_cost_type base_cost_type_;
+
+    vtr::interned_string empty_;
+
+    std::unordered_map<std::string, const t_physical_tile_type*> tile_type_to_pb_type_;
+
+    std::unordered_map<std::string, int> wire_name_to_seg_idx_;
+
+    std::set<uint32_t> bel_cell_mappings_;
+
+    std::unordered_map<int /*tile_id(name)*/, location> tile_to_loc_;
+    std::unordered_map<location, int /*tile_id(name)*/, hash_tuple::hash<location>> loc_to_tile_;
+
+    std::unordered_map<std::tuple<int /*timing_model*/, bool /*buffered*/>, int /*idx*/, hash_tuple::hash<std::tuple<int, bool>>> pips_models_;
+    std::unordered_map<std::tuple<int /*node_id*/, int /*node_id*/>, std::tuple<int /*switch_id*/, int /* tile_id */, std::tuple<int /*name*/, int /*wire0*/, int /*wire1*/, bool /*forward*/>>, hash_tuple::hash<std::tuple<int, int>>> pips_;
```

Comment:
I believe it may make sense to capture the big tuple in a struct, for clarity at least, as this is quite a big nested tuple

Reply:
In next commit I've added typedef from std::tuple<...> to pip_